Hid joystick interface

a-chol opened the pull request:
This implements a joystick feature, including a joystick_task function called from tmk, specific keycodes for joystick buttons and a usb hid interface for v-usb and lufa.
I tested the v-usb backend, as I have a compatible board, but only provide a compilable-but-untested implementation for lufa. Runtime tests of this backend is appreciated.
In order to test, you have to add JOYSTICK_ENABLE = yes to your rules.mk and 
```
#define JOYSTICK_BUTTON_COUNT 8
#define JOYSTICK_AXES_COUNT 2
```
in your config.h

Any feedback is appreciated.

---

closes #3903

mechmerlin commented on docs/feature_joystick.md:
```diff
@@ -0,0 +1,85 @@
+
+The keyboard can be made to be recognized as a joystick HID device by the Operating System. 
+
+This is enable by adding in your rule.mk : 
```
Just some wordsmithing

```suggestion
This is enabled by adding the following to `rules.mk`
```

mechmerlin commented on quantum/joystick.h:
```diff
@@ -0,0 +1,50 @@
+
+extern joystick_t joystick_status;
+
+//#define JS_INDEX(id, key) (key|((id & 0x7)<<6))
```
Can we just take out this commented out code?

a-chol replied:
You're right, I was initially considering supporting multiple instances of the interface, but decided it would be overreaching as a first PR.

mechmerlin commented on quantum/quantum_keycodes.h:
```diff
@@ -454,6 +454,42 @@ enum quantum_keycodes {
+    JS_BUTTON29,
+    JS_BUTTON30,
+    JS_BUTTON31,
+    JS_BUTTON_MAX = JS_BUTTON31,
```
Maybe add a 
```suggestion
JS_BUTTON_MIN = JS_BUTTON0
```

jackhumbert replied:
@mechmerlin it looks like your suggestion replaces that line - I think you can make one that adds a new line as well, right?

mechmerlin commented on tmk_core/protocol/lufa/lufa.c:
```diff
@@ -288,6 +292,84 @@ static void Console_Task(void)
+      return;
+    }
+    
+    joystick_report_t r = {
```
This code is very similar to the one in vusb.c, anyway to avoid some of the code repetition?

drashna commented:
It looks like this is added to VUSB and LUFA but not chibios, or arm_atsam.  Is there a reason why, other than a lack of hardware?

a-chol commented:
I'm AFK for a few days son I'll apply tour suggestions when I come back. 
I only implemented VUSB and Lufa mostly for hardware reasons. Giving Lufa a shot was a way to bring in more feedback and there is no reason to spare the others.

a-chol commented:
By the way, if anyone else has the hardware and feels like giving it a try, please.do. I'm not sure of the process of collaborating on a pull request on github though, maybe submitting a PR to my fork?

drashna commented:
Also, don't forget that if you're adding a feature page, you'll need to link it in /docs/_summary.md and /docs/_sidebar.md

coredump commented on docs/feature_joystick.md:
```diff
@@ -0,0 +1,85 @@
+In this example, the first axis will be read from the D7 pin, using an analogRead, whereas the second axis will not be read.
+In order to give a value to the second axis, you can do so in any custimizable entry point of quantum : as an action, in process_record_user or in matrix_scan_user, or even in joystick_task(void) which is call even when no key has been pressed.
+You assign a value by writing to joystick_status.axes[axis_index] a signed 8bit value (-127 to 127). Then it is necessary to assign the flag JS_UPDATED to joystick_status.status in order for the change to be taken into account.
+
```
Would like to have some clarification here. Is the `joystick_config_t` array necessary if you only want to use virtual joystick axis like described below, and if it is necessary, how to declare it since there is no actual pin with analog data?

a-chol replied:
Yes, the joystick_config_t is still necessary. I might be able to make it optional if all axes are virtual though. For now, the config would be as described in the doc, with all pin definitions defined as JS_VIRTUAL_AXIS. The two values after define the range of analog value read to map to -127;127, so there is no use for it for virtual axes.

a-chol replied:
@coredump I saw your post on reddit and that you are using a pro micro. Have you tried building a firmware with the joystick feature enabled yet?

coredump replied:
I did, it compiled, but I didn't add axis, just buttons. I got an error on Windows, like USB device not recognized. I saw some people on the Lufa github saying something about avoiding declaring a serial for the device (there's a env variable for that), but I haven't tested it yet.

coredump replied:
So, for some reason this is not compiling anymore. It's complaining about not finding the `joystick_config_t ` or anything defined on the quantum/* files related to the joystick support. I tried to find out why, but afaik the rules.mk is correct, and the common_features.mk should be adding the files to the compilation, but it's not happening.
There's a gist here with the error: https://gist.github.com/coredump/83b08aaab5e194f74785c2b94dd8c38d

(also I am running from your branch, @a-chol )

coredump replied:
I am doing the definition like this
```c
#ifdef JOYSTICK_ENABLE
joystick_config_t joystick_axes[JOYSTICK_AXES_COUNT] = {
  [0] = {JS_VIRTUAL_AXIS, JS_VIRTUAL_AXIS, 65280, 65472},
  [1] = {JS_VIRTUAL_AXIS, JS_VIRTUAL_AXIS, 65280, 65472},
  [2] = {JS_VIRTUAL_AXIS, JS_VIRTUAL_AXIS, 65280, 65472},
  [3] = {JS_VIRTUAL_AXIS, JS_VIRTUAL_AXIS, 65280, 65472},
  [4] = {JS_VIRTUAL_AXIS, JS_VIRTUAL_AXIS, 65280, 65472},
  [5] = {JS_VIRTUAL_AXIS, JS_VIRTUAL_AXIS, 65280, 65472},
  [6] = {JS_VIRTUAL_AXIS, JS_VIRTUAL_AXIS, 65280, 65472},
  [7] = {JS_VIRTUAL_AXIS, JS_VIRTUAL_AXIS, 65280, 65472},
};
#endif
```

The issue is that the files aren't being found or used by the compiler 

```console
Compiling: keyboards/fightkeyb/keymaps/default/keymap.c  
    keyboards/fightkeyb/keymaps/default/keymap.c:31:1: error: unknown type name 'joystick_config_t'; did you mean 'keymap_config_t'?
 joystick_config_t joystick_axes[JOYSTICK_AXES_COUNT] = {```

are you able to do a `make` successfully on your branch? I don't know why mine would fail.

a-chol replied:
Did you
#include <quantum/ joystick.h>
in your keymap.c?

coredump replied:
Ah. That was it.  is that on the doc somewhere?

Now I am able to compile and upload the firmware. My linux kernel is full of this:

```Dec 04 04:48:30 cammy kernel: hid-generic 0003:FEED:0000.002C: unknown main item tag 0x0
Dec 04 04:48:30 cammy kernel: hid-generic 0003:FEED:0000.002C: unknown main item tag 0x0
Dec 04 04:48:30 cammy kernel: hid-generic 0003:FEED:0000.002C: unknown main item tag 0x0
Dec 04 04:48:30 cammy kernel: hid-generic 0003:FEED:0000.002C: unknown main item tag 0x0
Dec 04 04:48:30 cammy kernel: hid-generic 0003:FEED:0000.002C: unknown main item tag 0x0
Dec 04 04:48:30 cammy kernel: hid-generic 0003:FEED:0000.002C: unknown main item tag 0x0```

And I can't see a joystick yet, but I will continue digging.

a-chol replied:
You're right, that's an oversight on my part, I'll add it. I'll check some more in the evening (CET) to see if I spot any mistake in the Lufa implementation. Do you have a tool that could display the hid report descriptors received for the device?

coredump replied:
https://gist.github.com/coredump/e924fb59d2b9301cf8e7f18b128641a1 I got some HID dumping done. I changed the `PRODUCT_ID` and `DEVICE_VERSION` on my config.h because it was conflicting with my other QMK keyboard on the same machine.

a-chol replied:
I spotted a few mistakes that could provide an incorrect device descriptor. Can you try to apply the following patch and see what changes?
https://gist.github.com/a-chol/97cbf2e1e1d7952324663459b817072c

a-chol replied:
Alright, that looks better. The only thing that seems off is the 
Report Count (7),
for the joysticks. Only 6 axes are supported, six usages defined just before so there is no seventh axis. Did you define JOYSTICK_AXES_COUNT 7 or 6?
Regarding the buttons, it looks good. There are 9 usages, 9 1-bit reports. You say that it reports the wrong number of buttons, did you not define JOYSTICK_BUTTON_COUNT to 9? What value did you give it?
Do the errors appear synchronously with you hitting buttons mapped to JS_BUTTON* keycodes?

coredump replied:
So, last message, I meant to say that it was reporting the wrong number of axis, not buttons, buttons are reported fine. I tried doing some changes to the report record to match some joysticks/gamepads I have here, so I ended with https://gist.github.com/coredump/f4cb72b5d6a0d2595d359527de0b0e37

I had the wrong number of axis on my config (I was using 7, sorry about that, off-by-1). I changed it to the default of 2 axis for testing now.

After that HID change, jstest started to show the correct number of axis(2).

Unfortunately, the USB-reset errors still happening and no input is seen by `evtest` or `jstest`. The resets don't happen in sync with the button press, it just keeps resetting as long as the testing programs are running, my guess is that the programs are trying to open the device and keep failing (that would explain the reset and the fact that I can't get any event coming from the joystick'. I don't think 

Interesting enough, since we have 2 report records (one for a keyboard, one for a joystick), I get two inputs on my kernel, and the keyboard input works fine (I have some buttons sending normal KC keycodes), it's just the joystick input that doesn't open/connect.

a-chol replied:
can you try applying the following patch?
https://gist.github.com/a-chol/3bdf405a65500c335148372edc8860a2
The send_joystick_packet function seems fine otherwise.

coredump replied:
That worked! At least on linux.
I got it to work and report the events to my js testing stuff. I rolled back all my changes to the HID descriptor too to check, and it seems to work, with one little caveat: if I use only 2 axis (0 and 1), this happens:
evtest = sees 2 axis
jstest = sees 0 axis
sdl2-jstest = sees 9 buttons and a ball (trackball)

So I added 6 axis back, and all of them report as axis then, for all the testing programs (so, linux input event subsystem and the joydev joystick driver).

Unfortunately, it appears as a broken USB device on windows, something about not being able to start the device, error `0xA` and `0xC000009C` that I searched and didn't get too far, as everything related to windows error codes. I did get some better error from a USB tool on widows and got `Report Descriptor was not byte aligned.` that sounds more likely to be a fixable bug.

But it is looking great!

a-chol replied:
Good new for linux! Can you provide me the diff of the changes you made, or offer a PR on my fork (if that's a thing, I'm not well versed in the usages of github). 
I use USBlyzer on windows, did you use that or does it give any more info. I don't really see how the report descriptor could be misaligned. 
Maybe you could try returning early from send_joystick_packet in lufa.c to avoid any report sending ; to check that it's not the report that is creating the error.
You could also vary the number of buttons : I suspect that using a number of buttons that is not a multiple of 8 wouldn't be handled well. I think I send the whole eight-bit array when I should only be sending the relevant bits. I'll look into it tomorrow.

coredump replied:
I forgot to let you know that I added a PR.

a-chol replied:
Excellent, it's now merged into this one! I'll add your fix for byte-aligned reports to the other existing implementations. Thanks for your help, hope you enjoy the feature!

drashna replied:
This got resolved, and added in, correct?

a-chol replied:
Yes, this got merged in my branch.

drashna replied:
Just adding some feedback here, it looks like you still need to manually add the include.  That's .... not good practice, and should be automatically included in "quantum.h" if the feature is enabled.

a-chol commented:
I just pushed an untested implementation of the hid joystick interface for chibios. It builds with the chibios_test keyboard, but lacks the analog read function. I couldn't find a simple function like the analogRead from the drivers/avr folder. It can still be usefull as is though.
So far, there is an implementation for AVR v-usb (tested), AVR LUFA (untested) and ARM chibios (untested). 
If anyone can give this a try on any of those untested platforms, it would be of great help.

a-chol commented:
I can give it a try but without the hardware to test, it's really hit-or-miss. I'm afraid that pushing incomplete implementation in this PR requires it to be validated before merging, which in the end postpones the merge for other protocols.
Unless we manage to have a tester for each protocol, I don't see that PR being merged anytime soon.

drashna commented:
Also, it looks like you have a merge conflict here.

drashna commented:
It looks like there are some merge conflicts that need to be resolved

a-chol commented:
The merge conflicts are resolved.
The feature is reaching completion. I've validated the adc reading code with my own keyboard, properly reading and mapping the inputs to the int8 value range.
Axis configuration now supports defining the input, output and ground pins to configure. When configured, the pin configuration is saved, then restored so that the qmk matrix scan, relying on input pins not changing, works properly.
Axis configuration also support defining calibration values using a low, resting and high raw adc reference values. Sometimes the resting position doesn't return the middle value of the extreme positions, but it is essential that the resting position return a value of 0 once mapped to the hid range, so a piecewise transfer function between raw adc values and the hid output range was required.
Beside these changes, I implemented the reading code in a function called from a weak linkage function, so that any keyboard can redefine the adc reading procedure and still call the original quantum code.

ADC reading is supported for AVR only for now. I think it would be more valuable for the project to merge it as is and rely on subsequent pull requests to implement the missing parts for other controllers.

sprayed commented:
Hi, I'm trying to write a jj4x4 keypad driver that makes the pad mimic a gamepad with only buttons. The hardware is very similar to the jj40 keyboard. Thanks to your branch I've been able to make this work! I've been able to adapt the jj40 settings to the jj4x4, but somehow under the game controller settings that now appear in windows, the game controller is still named jj40 even though i've renamed everything else (including the keyboard as it appears in control panel>devices and printers. I cannot find any references to jj40 in my code, so how does it still know it's a jj40? It's not essential since the code already works but i'm just curious. Again i'm extremely thankful to everyone working this feature!

drashna commented:
This is ready to go, correct?

a-chol commented:
I just pushed some documentation fixes, I confirm it's ready to go!

drashna commented:
There are some merge conflicts here.  If you could resolve these?

a-chol commented:
Conflicts are resolved. It's pretty cool to be able to do so from github directly!

drashna commented:
Unfortunately, some more merge conflicts have popped up!

drashna commented on common_features.mk:
```diff
@@ -295,8 +295,26 @@ ifeq ($(strip $(LEADER_ENABLE)), yes)
+
 include $(DRIVER_PATH)/qwiic/qwiic.mk
 
+ifeq ($(strip $(JOYSTICK_ENABLE)), yes)
```
Looks like this got added twice?

drashna replied:
This block looks to be identical, and added twice by accident.

One of them should be removed.

4sStylZ commented:
Hi @a-chol & @coredump, before trying to rebase and resolve the conflicts by myself I want to make a prototype working with the branch at this state. 

Do you have any example of a working keyboard to provide please ?

Thank you.

a-chol commented:
hello @4sStylZ, you can checkout my fork in the branch "farewell" which contains the farewell macropad. In its keymap folder, the "beko" keymap defines a dual axis joystick that I have tested an validated with a pro micro.
If you run into trouble resolving the conflict, hit me up so I can help.

4sStylZ commented:
@a-chol Thank you for the informations, I will check. 

Can you provide me some contact info as your github profile don’t show those info ? 
Like a discord / email / skype…

If you want you can send a reply to 4sStylZ#0812 on discord or 4sStylZ@protonmail.ch

Thank you.

drashna commented:
There are a bunch of merge conflicts that have popped up. 

Would you mind correcting these?

a-chol commented:
I fixed the conflicts and updated the code to keep in sync with some evolutions in quantum. It builds but I just need to find some time to test it on the hardware. I'll push to my fork when it's checked.

a-chol commented:
I pushed the merge into my fork.
Should I pay attention to the fact that checks are failing? I browsed the log but didn't see anything meaningfully related to my modifications.

yanfali commented:
@a-chol hey, I don't think you did the merge correctly. It might actually be easier to take a snapshot of your work and re-apply it on a clean checkout of master in a new branch. The problem is I think you changed all the commit hashes on the merge and basically created duplicate copies which have different hashes.

Let us know if you need some help, or jump on discord and we talk through some strategies for doing this correctly.

a-chol commented:
Damn, yeah, I didn't think of that. 
I used to rebase hid_joystick with master, but the merge was always painfull, as I had to resolve conflicts sometimes multiple times for the same chunk of code. So this time, I tried forking master and merging hid_joystick into it, but I didn't think later commits from master would be re-created.
Rebasing hid_joystick with master is still the only solution or is there a more convenient way to reapply master commits before mine in my branch?

4sStylZ commented:
Maybe squash all the commits of your branch on your repo for having the minimum change on the same commit.
After that, rebasing the upstream master qmk branch.
You will have conflicts but only one time.

But yeah the discord is the best way to speak about that

a-chol commented:
Ok, I redid the merge the proper way and squashed all the commits. This is good to go.

drashna commented on drivers/avr/analog.c:
```diff
@@ -21,17 +21,55 @@
 
 void analogReference(uint8_t mode) { aref = mode & 0xC0; }
 
+uint8_t pinToADCIndex(uint8_t pin){
```
It looks like this change is breaking something in `convert/usb_usb/ble`.

drashna replied:
Specifically, BLE also uses analog.c, but there is some sort of incompatibility here.  Mostly, because I think that the board in question doesn't have the config_common.h file included that includes these definitions. 
```
Compiling: drivers/avr/analog.c                                                                    drivers/avr/analog.c: In function 'pinToADCIndex':
drivers/avr/analog.c:31:7: error: 'B4' undeclared (first use in this function); did you mean 'PB4'?
       B4, 8,
       ^~
       PB4
drivers/avr/analog.c:31:7: note: each undeclared identifier is reported only once for each function it appears in
drivers/avr/analog.c:32:7: error: 'B5' undeclared (first use in this function); did you mean 'PB5'?
       B5, 7,
       ^~
       PB5
```

a-chol replied:
I can't find where the BLE keyboard calls or builds the analog file. I should be able to reproduce the problem with "make converter/usb_usb/ble"?

drashna commented on quantum/quantum_keycodes.h:
```diff
@@ -503,6 +503,43 @@ enum quantum_keycodes {
     MAGIC_UNSWAP_CTL_GUI,
     MAGIC_TOGGLE_CTL_GUI,
 
+#ifdef JOYSTICK_ENABLE
```
Lets not hid the buttons behind if-defs.  This way, they'll compile fine, but just not work.  
```suggestion
```

a-chol replied:
change integrated in the last merge

drashna commented on quantum/quantum_keycodes.h:
```diff
@@ -503,6 +503,43 @@ enum quantum_keycodes {
+    JS_BUTTON30,
+    JS_BUTTON31,
+    JS_BUTTON_MAX = JS_BUTTON31,
+#endif
```
```suggestion
```

a-chol replied:
change integrated in the last merge

drashna commented:
It looks like there are some merge conflicts that have popped up here.

also, it looks like there are some issues with analog.c, depending on the hardware.

drashna commented:
Just tested it, and can definitely verify that it works.

RJMakesStuff commented:
I'm new to Github so I don't really get forking or anything but I rigged a hand wired board with a 3x4 keypad and joystick using your changes and a custom 32u4 dev board. The only problem I've had is JS_BUTTON0 etc. go to random buttons in windows. Button 0 is 5 for me.

drashna commented:
There are a number of merge conflicts here.

drashna commented:
@a-chol any chance that you could fix the merge conflicts here?

a-chol commented:
I just fixed the conflicts and tested the merge, all is well.

tzarc commented:
Added a PR for your branch to remove the outstanding duplicate.

tzarc commented:
Didn't see @drashna's other requested changes... another PR added.

tzarc commented:
Also added a PR for a onekey with a joystick button.

Getting the following when attempting to get it loaded up:
![image](https://user-images.githubusercontent.com/2985843/72100023-f69e9680-3375-11ea-9d45-d3b5ed8e4190.png)

tzarc commented:
Okay, diagnosed the issue -- the USB descriptor wasn't capable of dealing with a zero for either axes or buttons. Added the fix to the onekey PR, getting this now:

![image](https://user-images.githubusercontent.com/2985843/72101526-0370b980-3379-11ea-828c-47d42f2e6217.png)

drashna commented:
> Okay, diagnosed the issue -- the USB descriptor wasn't capable of dealing with a zero for either axes or buttons. Added the fix to the onekey PR, getting this now:

If that's the case, maybe we should generate an error if that is the case, if possible.

tzarc commented:
> 
> 
> > Okay, diagnosed the issue -- the USB descriptor wasn't capable of dealing with a zero for either axes or buttons. Added the fix to the onekey PR, getting this now:
> 
> If that's the case, maybe we should generate an error if that is the case, if possible.

Yep, added to the PR for their branch, will hopefully filter through.
Added a compile-time check that there's at least one axis or button as well.

tzarc commented:
Bugger, conflicts again.

tzarc commented:
Tested the onekey on both Pro Micro and Proton C.

zvecr commented:
@fauxpark any chance you could have a look at the USB endpoint stuff?

fauxpark commented on tmk_core/protocol/lufa/lufa.c:
```diff
@@ -268,6 +272,85 @@ static void Console_Task(void) {
+    #if JOYSTICK_BUTTON_COUNT>0
+    uint8_t buttons[(JOYSTICK_BUTTON_COUNT-1)/8+1];
+    #endif
+} __attribute__ ((packed)) joystick_report_t;
```
This should go in report.h.

fauxpark commented on tmk_core/protocol/usb_descriptor.c:
```diff
@@ -287,7 +345,6 @@ const USB_Descriptor_Device_t PROGMEM DeviceDescriptor = {
     },
     .USBSpecification           = VERSION_BCD(1, 1, 0),
-
 #if VIRTSER_ENABLE
```
```suggestion

#if VIRTSER_ENABLE
```

fauxpark commented on tmk_core/protocol/usb_descriptor.c:
```diff
@@ -812,6 +869,49 @@ const USB_Descriptor_Configuration_t PROGMEM ConfigurationDescriptor = {
     },
 #endif
+
+
```
```suggestion
```

fauxpark commented on tmk_core/protocol/usb_descriptor.c:
```diff
@@ -812,6 +869,49 @@ const USB_Descriptor_Configuration_t PROGMEM ConfigurationDescriptor = {
+            .EndpointAddress        = (ENDPOINT_DIR_IN | JOYSTICK_IN_EPNUM),
+            .Attributes             = (EP_TYPE_INTERRUPT | ENDPOINT_ATTR_NO_SYNC | ENDPOINT_USAGE_DATA),
+            .EndpointSize           = JOYSTICK_EPSIZE,
+            .PollingIntervalMS      = 0x0A
```
```suggestion
            .PollingIntervalMS      = USB_POLLING_INTERVAL_MS
```

fauxpark commented on drivers/avr/analog.c:
```diff
@@ -128,5 +128,10 @@ int16_t adc_read(uint8_t mux) {
+    low |= (ADCH << 8);
+    
+    //turn off the ADC
+    ADCSRA &= ~(1<<ADEN);
```
Why does this need to be done?

a-chol replied:
The pin being used as ADC here could be a pin used in the keyboard matrix (hence the saveState/restoreState in process_joystick) so leaving the ADC on could interfere. I'm aware firing up the ADC every time might be a bit costly, how much of a problem do you think this would be?

fauxpark replied:
If the pin is already being used in the matrix, it should not be read by the ADC, as the matrix scan is happening continuously and will interfere.

a-chol replied:
As far as I understand, the process call of each feature is sequential, si the matrix scan happens completely, then other features are executed. This is not parallel so there shouldn't be any problem reusing the pins from the matrix to read some analog input. In fact, this is what I did on my keyboard (BminiEx with ATMega32A) and it works fine, given a proper use of diodes to isolate the matrix scanning circuit from the analog reading circuit.

fauxpark commented on quantum/joystick.h:
```diff
@@ -0,0 +1,55 @@
+#ifndef JOYSTICK_H
```
```suggestion
#pragma once
```

fauxpark commented on quantum/joystick.h:
```diff
@@ -0,0 +1,55 @@
+#ifndef JOYSTICK_H
+#define JOYSTICK_H
```
```suggestion
```

fauxpark commented on quantum/joystick.h:
```diff
@@ -0,0 +1,55 @@
+//to be implemented in the hid protocol library
+void send_joystick_packet(joystick_t* joystick);
+
+#endif //JOYSTICK_H
```
```suggestion
```

fauxpark commented on quantum/process_keycode/process_joystick.c:
```diff
@@ -0,0 +1,160 @@
+#include "process_joystick.h"
+
+#include <quantum/joystick.h>
+#include <quantum/quantum_keycodes.h>
```
```suggestion
```

fauxpark commented on quantum/process_keycode/process_joystick.c:
```diff
@@ -0,0 +1,160 @@
+#include "process_joystick.h"
+
+#include <quantum/joystick.h>
```
```suggestion
```

fauxpark commented on quantum/process_keycode/process_joystick.c:
```diff
@@ -0,0 +1,160 @@
+
+#include <quantum/joystick.h>
+#include <quantum/quantum_keycodes.h>
+#include <quantum/config_common.h>
```
```suggestion
```

fauxpark commented on quantum/process_keycode/process_joystick.c:
```diff
@@ -0,0 +1,160 @@
+#include <quantum/config_common.h>
+
+#ifdef __AVR__
+# include <drivers/avr/analog.h>
```
```suggestion
#    include "analog.h"
```

fauxpark commented on quantum/process_keycode/process_joystick.c:
```diff
@@ -0,0 +1,160 @@
+#else
+  return;
+#endif
+}
```
Why not just use `readPin()`, `writePin*()`?

a-chol replied:
readPin reads the state of input pins and writePin writes output pins, whereas here we need to read the state of output pins and data direction to be able to restore later.

fauxpark commented on quantum/process_keycode/process_joystick.c:
```diff
@@ -0,0 +1,160 @@
+    //test the converted value against the lower range
+    uint16_t ref = joystick_axes[axis_index].mid_digit;
+    uint16_t range = joystick_axes[axis_index].min_digit;
+    int16_t ranged_val = -127*fminf(1.f, (axis_val - (float)(ref)) /(range - (float)ref));
```
Is there any way we can avoid using floats? This will greatly increase the firmware size.

a-chol replied:
I'll look into it

fauxpark commented on quantum/process_keycode/process_joystick.h:
```diff
@@ -0,0 +1,14 @@
+#ifndef PROCESS_JOYSTICK_H
```
```suggestion
#pragma once
```

fauxpark commented on quantum/process_keycode/process_joystick.h:
```diff
@@ -0,0 +1,14 @@
+#ifndef PROCESS_JOYSTICK_H
+#define PROCESS_JOYSTICK_H
```
```suggestion
```

fauxpark commented on quantum/process_keycode/process_joystick.h:
```diff
@@ -0,0 +1,14 @@
+bool process_joystick_analogread(void);
+bool process_joystick_analogread_quantum(void);
+
+#endif //PROCESS_JOYSTICK_H
```
```suggestion
```

fauxpark commented on quantum/quantum.h:
```diff
@@ -138,6 +138,10 @@ extern layer_state_t layer_state;
 #endif
 
+#ifdef JOYSTICK_ENABLE
+#    include "joystick.h"
```
```suggestion
#    include "process_joystick.h"
```

fauxpark commented on quantum/quantum.c:
```diff
@@ -62,6 +62,10 @@ float bell_song[][2] = SONG(TERMINAL_SOUND);
 #    endif
 #endif
 
+#ifdef JOYSTICK_ENABLE
```
```suggestion
```

fauxpark commented on quantum/quantum.c:
```diff
@@ -62,6 +62,10 @@ float bell_song[][2] = SONG(TERMINAL_SOUND);
 #endif
 
+#ifdef JOYSTICK_ENABLE
+#include <process_keycode/process_joystick.h>
```
```suggestion
```

fauxpark commented on quantum/quantum.c:
```diff
@@ -62,6 +62,10 @@ float bell_song[][2] = SONG(TERMINAL_SOUND);
 
+#ifdef JOYSTICK_ENABLE
+#include <process_keycode/process_joystick.h>
+#endif //JOYSTICK_ENABLE
```
```suggestion
```

fauxpark commented on quantum/quantum.c:
```diff
@@ -62,6 +62,10 @@ float bell_song[][2] = SONG(TERMINAL_SOUND);
+#ifdef JOYSTICK_ENABLE
+#include <process_keycode/process_joystick.h>
+#endif //JOYSTICK_ENABLE
+
```
```suggestion
```

fauxpark commented on tmk_core/protocol/chibios/usb_main.c:
```diff
@@ -47,6 +47,10 @@
 #endif
 
+#ifdef JOYSTICK_ENABLE
+# include <quantum/joystick.h>
```
```suggestion
#    include "joystick.h"
```

fauxpark commented on tmk_core/protocol/vusb/vusb.c:
```diff
@@ -26,6 +26,8 @@ along with this program.  If not, see <http://www.gnu.org/licenses/>.
 #include "host_driver.h"
 #include "vusb.h"
+#include "joystick.h"
+#include "joystick.h"
```
```suggestion
```

tzarc commented:
I'll run thru @fauxpark's requests and get another PR across to @a-chol tomorrow.

tzarc commented:
So I've got an updated version, rebased on top of latest master, with the majority of the ewview comments here:
https://github.com/tzarc/qmk_firmware/tree/hid_joystick

@a-chol do you want to grab it and merge into your local repo?
There are a couple of things still outstanding, like the disabling of the ADC after use, and whether or not we can use integer arithmetic instead -- it'd be great if you could have a bit of a look at whether they're possible.

tzarc commented:
As far as I can tell, my branch isn't going to work with the shared endpoint stuff, specifically when we're looking at mouse/extra/joystick reports getting merged in `usb_descriptor.c`.

tzarc commented:
@a-chol, out of curiosity -- do you want to continue this PR, or do you think it'd be easier if I opened a new one and fixed up the discrepancies to push it through?

a-chol commented:
Hey, I'll review your changes and merge them, sorry for the delay.Le 18 mars 2020 03:00, Nick Brassel <notifications@github.com> a écrit :
@a-chol, out of curiosity -- do you want to continue this PR, or do you think it'd be easier if I opened a new one and fixed up the discrepancies to push it through?

—You are receiving this because you were mentioned.Reply to this email directly, view it on GitHub, or unsubscribe.

a-chol commented on quantum/process_keycode/process_joystick.c:
```diff
@@ -0,0 +1,160 @@
+  return true;
+}
+
+__attribute__ ((weak))
```
Anytime the user wants to handle joystick in a different way. A couple usecases : joystick inputs are simulated based on keyboard input only. Joystick inputs are read but won't be sent as a hid report (to move the mouse instead, or change volume).

a-chol commented on docs/feature_joystick.md:
```diff
@@ -0,0 +1,145 @@
+
+### Analog circuit
+
+An analog device such as a potentiometer found on a gamepad's analog axes is based on a [voltage divider](https://en.wikipedia.org/wiki/Voltage_divider).
```
Agreed

a-chol commented:
@tzarc Thanks for your work putting in the review changes.
I'm not sure why V-USB has to be taken out though. Do you lack testing equipment or is there a deeper reason I missed?

fauxpark commented:
@a-chol there's some work I want to do on the V-USB code first in order to make adding things like this, console and raw HID easier and more consistent across V-USB/LUFA/ChibiOS/etc. As it is, just tacking the joystick descriptors onto the existing interface is... not ideal.

a-chol commented:
@fauxpark I agree it's not ideal but I didn't manage to get it working another way. I am no expert of the USB standard, but I think you need to have more endpoints declared if you want to separate things, and it didn't look like V-USB allowed declaring more endpoints. I'd be happy to help if you have more insights in the standard or V-USB.

fauxpark commented:
> it didn't look like V-USB allowed declaring more endpoints

I have some ideas on how to do that. It'll just be easier if I only have to deal with the existing stuff.

I'm also not sure how useful joystick support will be on V-USB. The boards that use it are mainly existing PS2AVRGB boards, and this code looks fairly hefty (not your fault though, just seems to be how it is). The only two chips QMK currently supports with V-USB only have 32kB of flash space; even less when you factor in the bootloader.
So IMO the Chibios/LUFA support should be the main focus as most new boards are going to be using either ARM chips, or the USB AVRs like the 32U4.

a-chol commented:
I've cleaned up the history of the PR which was a mess by my own fault and made rebasing master a PITA. 
@tzarc I've taken the liberty to squash most of your commits if you don't mind.

@fauxpark  Yeaaah buuut... I'm using it on a ps2avrgb board :sweat:
I mean, I don't necessarily have to have it on the PR branch, I can keep the current implementation for vusb local, but it feels like a waste to have it and not share it.

tzarc commented:
Yeah no problem at all on the squash.
We've also picked up ADC on ARM -- don't worry about getting that into the PR, I'll sort it out when we get this stuff in.

tzarc commented:
As for V-USB -- @fauxpark can confirm what we'd like to do here.
As stated there's some work on figuring out how to deal with it in general -- with mousekeys, extrakey, and now joystick... the current functionality to deal with merging endpoints isn't as straightforward.

a-chol commented:
Allright, I pushed your commit into the PR branch, I can still revert it on demand when I need to build the firmware for my keyboard.

fauxpark commented:
@a-chol Touché 😄

To be clear, I'm not against adding joystick support for V-USB. I'd at least prefer it to be in a separate PR though, just due to the other stuff going on with the V-USB codebase.

fauxpark commented on tmk_core/protocol/usb_descriptor.c:
```diff
@@ -278,6 +278,64 @@ const USB_Descriptor_HIDReport_Datatype_t PROGMEM ConsoleReport[] = {
+#ifdef JOYSTICK_ENABLE
+#if JOYSTICK_AXES_COUNT == 0 && JOYSTICK_BUTTON_COUNT == 0
+    #error Need at least one axis or button for joystick
+#endif
```
```suggestion
#    if JOYSTICK_AXES_COUNT == 0 && JOYSTICK_BUTTON_COUNT == 0
#        error Need at least one axis or button for joystick
#    endif
```

fauxpark commented on tmk_core/protocol/usb_descriptor.c:
```diff
@@ -278,6 +278,64 @@ const USB_Descriptor_HIDReport_Datatype_t PROGMEM ConsoleReport[] = {
+                #endif
+            #endif
+        HID_RI_END_COLLECTION(0),
+    HID_RI_END_COLLECTION(0),
```
```suggestion
const USB_Descriptor_HIDReport_Datatype_t PROGMEM JoystickReport[] = {
    HID_RI_USAGE_PAGE(8, 0x01),         // Generic Desktop
    HID_RI_USAGE(8, 0x04),              // Joystick
    HID_RI_COLLECTION(8, 0x01),         // Application
        HID_RI_COLLECTION(8, 0x00),     // Physical
            HID_RI_USAGE_PAGE(8, 0x01), // Generic Desktop
#    if JOYSTICK_AXES_COUNT >= 1
            HID_RI_USAGE(8, 0x30),      // X
#    endif
#    if JOYSTICK_AXES_COUNT >= 2
            HID_RI_USAGE(8, 0x31),      // Y
#    endif
#    if JOYSTICK_AXES_COUNT >= 3
            HID_RI_USAGE(8, 0x32),      // Z
#    endif
#    if JOYSTICK_AXES_COUNT >= 4
            HID_RI_USAGE(8, 0x33),      // Rx
#    endif
#    if JOYSTICK_AXES_COUNT >= 5
            HID_RI_USAGE(8, 0x34),      // Ry
#    endif
#    if JOYSTICK_AXES_COUNT >= 6
            HID_RI_USAGE(8, 0x35),      // Rz
#    endif
#    if JOYSTICK_AXES_COUNT >= 1
            HID_RI_LOGICAL_MINIMUM(8, -127),
            HID_RI_LOGICAL_MAXIMUM(8, 127),
            HID_RI_REPORT_COUNT(8, JOYSTICK_AXES_COUNT),
            HID_RI_REPORT_SIZE(8, 0x08),
            HID_RI_INPUT(8, HID_IOF_DATA | HID_IOF_VARIABLE | HID_IOF_ABSOLUTE),
#    endif

#    if JOYSTICK_BUTTON_COUNT >= 1
            HID_RI_USAGE_PAGE(8, 0x09), // Button
            HID_RI_USAGE_MINIMUM(8, 0x01),
            HID_RI_USAGE_MAXIMUM(8, JOYSTICK_BUTTON_COUNT),
            HID_RI_LOGICAL_MINIMUM(8, 0x00),
            HID_RI_LOGICAL_MAXIMUM(8, 0x01),
            HID_RI_REPORT_COUNT(8, JOYSTICK_BUTTON_COUNT),
            HID_RI_REPORT_SIZE(8, 0x01),
            HID_RI_INPUT(8, HID_IOF_DATA | HID_IOF_VARIABLE | HID_IOF_ABSOLUTE),

#        if (JOYSTICK_BUTTON_COUNT % 8) != 0
            HID_RI_REPORT_COUNT(8, 8 - (JOYSTICK_BUTTON_COUNT % 8)),
            HID_RI_REPORT_SIZE(8, 0x01),
            HID_RI_INPUT(8, HID_IOF_CONSTANT),
#        endif
#    endif
        HID_RI_END_COLLECTION(0),
    HID_RI_END_COLLECTION(0)
```

fauxpark commented on tmk_core/protocol/usb_descriptor.h:
```diff
@@ -224,6 +234,10 @@ enum usb_endpoints {
 #endif
+#ifdef JOYSTICK_ENABLE
+    JOYSTICK_IN_EPNUM  = NEXT_EPNUM,
+    JOYSTICK_OUT_EPNUM = NEXT_EPNUM,
```
```suggestion
```

a-chol replied:
I merged your changes removing the output capacity on the joystick endpoint, but it seems to create a build error : 
tmk_core/protocol/chibios/usb_main.c:297:46: error: 'JOYSTICK_OUT_CAPACITY' undeclared here (not in a function); did you mean 'JOYSTICK_IN_CAPACITY'?
     .joystick_driver = QMK_USB_DRIVER_CONFIG(JOYSTICK, 0, false),
                                              ^~~~~~~~
tmk_core/protocol/chibios/usb_main.c:184:74: note: in definition of macro 'QMK_USB_DRIVER_CONFIG'
         .queue_capacity_in = stream##_IN_CAPACITY, .queue_capacity_out = stream##_OUT_CAPACITY,

It seems the output capacity is expected by the macro.
Either I don't use the macro to define the endpoint and do it manually (or with a macro that defines only the input capacity), if that's possible, or I revert your changes to keep the output capacity although it won't be used. Which one to you prefer?

fauxpark replied:
Ah yeah...oops. The out mode and capacity should go back as they were (even though there is only device-to-host communication for the joystick endpoint).

fauxpark commented on tmk_core/protocol/chibios/usb_main.c:
```diff
@@ -283,6 +290,14 @@ static usb_driver_configs_t drivers = {
+
+#ifdef JOYSTICK_ENABLE
+#    define JOYSTICK_IN_CAPACITY 4
+#    define JOYSTICK_OUT_CAPACITY 4
```
```suggestion
```

fauxpark commented on tmk_core/protocol/chibios/usb_main.c:
```diff
@@ -283,6 +290,14 @@ static usb_driver_configs_t drivers = {
+#    define JOYSTICK_IN_CAPACITY 4
+#    define JOYSTICK_OUT_CAPACITY 4
+#    define JOYSTICK_IN_MODE USB_EP_MODE_TYPE_BULK
+#    define JOYSTICK_OUT_MODE USB_EP_MODE_TYPE_BULK
```
```suggestion
```

a-chol commented:
> 
> 
> @a-chol Touché 😄
> 
> To be clear, I'm not against adding joystick support for V-USB. I'd at least prefer it to be in a separate PR though, just due to the other stuff going on with the V-USB codebase.

No worries, I wasn't aware there was work done on the V-USB implementation on another PR. It makes sense to avoid concurrent modifications.

a-chol commented on common_features.mk:
```diff
@@ -511,3 +511,12 @@ ifeq ($(strip $(AUTO_SHIFT_ENABLE)), yes)
+    SRC += $(QUANTUM_DIR)/joystick.c
+    ifeq ($(PLATFORM),AVR)
+      SRC += drivers/avr/analog.c
+    endif
```
I'm looking into it, I bought a proton C a few month ago so I should be able to port and test.

a-chol replied:
I've finished porting and fixing the joystick specific code for chibios-backed MCUs. It's tested with a proton C. I would appreciate testing on other MCU if anyone have some available. I added a joystick axis to the onekey:joystick keymap. I'll look into adding a second axis that animates without requiring an analog circuitry to validate the HID protocol part.

fauxpark commented on quantum/process_keycode/process_joystick.c:
```diff
@@ -0,0 +1,168 @@
+}
+
+uint16_t savePinState(uint32_t pin) {
+#ifdef __AVR__
```
```suggestion
uint16_t savePinState(pin_t pin) {
#if defined(PROTOCOL_LUFA)
```

fauxpark commented on quantum/process_keycode/process_joystick.c:
```diff
@@ -0,0 +1,168 @@
+}
+
+void restorePinState(uint32_t pin, uint16_t restoreState) {
+#ifdef __AVR__
```
```suggestion
void restorePinState(pin_t pin, uint16_t restoreState) {
#if defined(PROTOCOL_LUFA)
```

fauxpark commented on quantum/joystick.h:
```diff
@@ -0,0 +1,52 @@
+typedef struct {
+    uint32_t output_pin;
+    uint32_t input_pin;
+    uint32_t ground_pin;
```
```suggestion
    pin_t output_pin;
    pin_t input_pin;
    pin_t ground_pin;
```

fauxpark commented on tmk_core/protocol/chibios/usb_main.c:
```diff
@@ -870,3 +885,65 @@ void virtser_task(void) {
+#        endif
+#        if JOYSTICK_BUTTON_COUNT > 24
+                    ,
+                    joystick->buttons[3]
```
```suggestion
        .axes = {joystick->axes[0],

#        if JOYSTICK_AXES_COUNT >= 2
                 joystick->axes[1],
#        endif
#        if JOYSTICK_AXES_COUNT >= 3
                 joystick->axes[2],
#        endif
#        if JOYSTICK_AXES_COUNT >= 4
                 joystick->axes[3],
#        endif
#        if JOYSTICK_AXES_COUNT >= 5
                 joystick->axes[4],
#        endif
#        if JOYSTICK_AXES_COUNT >= 6
                 joystick->axes[5],
#        endif
        },
#    endif  // JOYSTICK_AXES_COUNT>0

#    if JOYSTICK_BUTTON_COUNT > 0
        .buttons = {joystick->buttons[0],

#        if JOYSTICK_BUTTON_COUNT > 8
                    joystick->buttons[1],
#        endif
#        if JOYSTICK_BUTTON_COUNT > 16
                    joystick->buttons[2],
#        endif
#        if JOYSTICK_BUTTON_COUNT > 24
                    joystick->buttons[3],
```

fauxpark commented on tmk_core/protocol/lufa/lufa.c:
```diff
@@ -263,6 +267,74 @@ static void Console_Task(void) {
+#        endif
+#        if JOYSTICK_BUTTON_COUNT > 24
+                    ,
+                    joystick->buttons[3]
```
```suggestion
        .axes = {joystick->axes[0],

#        if JOYSTICK_AXES_COUNT >= 2
                 joystick->axes[1],
#        endif
#        if JOYSTICK_AXES_COUNT >= 3
                 joystick->axes[2],
#        endif
#        if JOYSTICK_AXES_COUNT >= 4
                 joystick->axes[3],
#        endif
#        if JOYSTICK_AXES_COUNT >= 5
                 joystick->axes[4],
#        endif
#        if JOYSTICK_AXES_COUNT >= 6
                 joystick->axes[5],
#        endif
        },
#    endif  // JOYSTICK_AXES_COUNT>0

#    if JOYSTICK_BUTTON_COUNT > 0
        .buttons = {joystick->buttons[0],

#        if JOYSTICK_BUTTON_COUNT > 8
                    joystick->buttons[1],
#        endif
#        if JOYSTICK_BUTTON_COUNT > 16
                    joystick->buttons[2],
#        endif
#        if JOYSTICK_BUTTON_COUNT > 24
                    joystick->buttons[3],
```

a-chol commented on docs/feature_joystick.md:
```diff
@@ -0,0 +1,149 @@
+Both services can be used without the other, depending on whether you just want to read a device but not send gamepad reports (for volume control for instance)
+or send gamepad reports based on values computed by the keyboard.
+
+!> Reading analog axes is not currently functional on ARM MCUs.
```
This is tested on a proton c already. You can try it for confirmation if you'd like. With the onekey firmware, the first axis is read from A0.

a-chol commented on docs/feature_joystick.md:
```diff
@@ -0,0 +1,149 @@
+### Configuring the joystick
+
+The default joystick has 2 axes and and 8 buttons. This can be changed from the config.h file :
+
```
Yeah, it seemed like a sensible default to have axes and buttons to match a usual joystick.

a-chol commented on quantum/process_keycode/process_joystick.c:
```diff
@@ -0,0 +1,150 @@
+#else
+    return;
+#endif
+}
```
My last modifications implement the behaviour for chibios, but that's not what I would call "platform-agnosticism". Would you prefer me to implement save and restore functions inside the respective driver folder of each platform to be able have the same function call inside the quantum code?

drashna commented:
Looks good!  I would prefer that it was not dependant on analog, but I think that separating that out may be best for a separate PR (when somebody implementents a different input method).  

I'd prefer of @fauxpark signs off on this first, though

zvecr commented on keyboards/handwired/onekey/keymaps/joystick/halconf.h:
```diff
@@ -0,0 +1,524 @@
+/*
```
This file should really be removed, given it cannot apply to all chibios chips.

zvecr commented on keyboards/handwired/onekey/keymaps/joystick/mcuconf.h:
```diff
@@ -0,0 +1,273 @@
+/*
```
same for this file, should be removed.

zvecr commented on keyboards/handwired/onekey/proton_c/config.h:
```diff
@@ -29,4 +29,4 @@
 #define RGB_DI_PIN A0
 
-#define ADC_PIN A0
+#define ADC_PIN A1
```
```suggestion
#define ADC_PIN A0
```

drashna replied:
Doesn't this make it share the pin with rgblight, @zvecr?

fauxpark replied:
I think the idea with the onekey boards is that although both are using the same pin, under normal circumstances only one of these is enabled at a time, depending on the keymap built.

a-chol replied:
I made the change to A0, no worries.

a-chol commented:
All requested changes have been committed.

tzarc commented:
Epic struggle, but we got there in the end. Cheers!

sigprof commented on tmk_core/protocol/lufa/lufa.c:
```diff
@@ -410,6 +474,9 @@ void EVENT_USB_Device_ConfigurationChanged(void) {
     ConfigSuccess &= Endpoint_ConfigureEndpoint(CDC_IN_EPADDR, EP_TYPE_BULK, CDC_EPSIZE, ENDPOINT_BANK_SINGLE);
 #endif
+#ifdef JOYSTICK_ENABLE
+    ConfigSuccess &= ENDPOINT_CONFIG(JOYSTICK_IN_EPNUM, EP_TYPE_INTERRUPT, ENDPOINT_DIR_IN, JOYSTICK_EPSIZE, ENDPOINT_BANK_SINGLE);
```
This does not compile in the `develop` branch — the `ENDPOINT_CONFIG` macro was removed, and the code needs to be changed to:
```
    ConfigSuccess &= Endpoint_ConfigureEndpoint((JOYSTICK_IN_EPNUM | ENDPOINT_DIR_IN), EP_TYPE_INTERRUPT, JOYSTICK_EPSIZE, 1);
```

sigprof commented on keyboards/handwired/onekey/keymaps/joystick/keymap.c:
```diff
@@ -0,0 +1,25 @@
+#endif
+
+const uint16_t PROGMEM keymaps[][MATRIX_ROWS][MATRIX_COLS] = {
+  LAYOUT( JS_BUTTON0 )
```
This does not compile after some recent changes to `handwired/onekey` — the `LAYOUT` macro was renamed:
```c
  LAYOUT_ortho_1x1( JS_BUTTON0 )
```

sigprof commented on quantum/process_keycode/process_joystick.c:
```diff
@@ -0,0 +1,168 @@
+    uint8_t pinNumber = pin & 0xF;
+    return ((PORTx_ADDRESS(pin) >> pinNumber) & 0x1) << 1 | ((DDRx_ADDRESS(pin) >> pinNumber) & 0x1);
+#elif defined(PROTOCOL_CHIBIOS)
+    /*
+    The pin configuration is backed up in the following format : 
+ bit  15    9  8   7   6  5  4   3     2    1 0
+      |unused|ODR|IDR|PUPDR|OSPEEDR|OTYPER|MODER|
+    */
+    return  (( PAL_PORT(pin)->MODER   >> (2*PAL_PAD(pin))) & 0x3)
+          | (((PAL_PORT(pin)->OTYPER  >> (1*PAL_PAD(pin))) & 0x1) << 2) 
+          | (((PAL_PORT(pin)->OSPEEDR >> (2*PAL_PAD(pin))) & 0x3) << 3)
+          | (((PAL_PORT(pin)->PUPDR   >> (2*PAL_PAD(pin))) & 0x3) << 5)
+          | (((PAL_PORT(pin)->IDR     >> (1*PAL_PAD(pin))) & 0x1) << 7)
+          | (((PAL_PORT(pin)->ODR     >> (1*PAL_PAD(pin))) & 0x1) << 8);
```
This code is problematic — it is not just ChibiOS-specific, but works only with some particular STM32 chips (e.g., it does not compile for `handwired/onekey/bluepill:joystick`, which has STM32F103 with a different GPIO hardware block). As a minimum, it should have some hardware-specific `#ifdef`s, so that unsupported chips could still support a virtual joystick.

sigprof commented on quantum/process_keycode/process_joystick.c:
```diff
@@ -0,0 +1,168 @@
+    PORTx_ADDRESS(pin) = (PORTx_ADDRESS(pin) & ~_BV(pinNumber)) | (((restoreState >> 1) & 0x1) << pinNumber);
+    DDRx_ADDRESS(pin)  = (DDRx_ADDRESS(pin) & ~_BV(pinNumber)) | ((restoreState & 0x1) << pinNumber);
+#elif defined(PROTOCOL_CHIBIOS)
+    PAL_PORT(pin)->MODER  =  (PAL_PORT(pin)->MODER   & ~(0x3<< (2*PAL_PAD(pin)))) | (restoreState & 0x3)     << (2*PAL_PAD(pin));
+    PAL_PORT(pin)->OTYPER =  (PAL_PORT(pin)->OTYPER  & ~(0x1<< (1*PAL_PAD(pin)))) | ((restoreState>>2) & 0x1) << (1*PAL_PAD(pin));
+    PAL_PORT(pin)->OSPEEDR=  (PAL_PORT(pin)->OSPEEDR & ~(0x3<< (2*PAL_PAD(pin)))) | ((restoreState>>3) & 0x3) << (2*PAL_PAD(pin));
+    PAL_PORT(pin)->PUPDR  =  (PAL_PORT(pin)->PUPDR   & ~(0x3<< (2*PAL_PAD(pin)))) | ((restoreState>>5) & 0x3) << (2*PAL_PAD(pin));
+    PAL_PORT(pin)->IDR    =  (PAL_PORT(pin)->IDR     & ~(0x1<< (1*PAL_PAD(pin)))) | ((restoreState>>7) & 0x1) << (1*PAL_PAD(pin));
+    PAL_PORT(pin)->ODR    =  (PAL_PORT(pin)->ODR     & ~(0x1<< (1*PAL_PAD(pin)))) | ((restoreState>>8) & 0x1) << (1*PAL_PAD(pin));
```
Another part of the hardware-specific code which should be guarded with more specific `#ifdef`s than just `defined(PROTOCOL_CHIBIOS)`.

4sStylZ commented:
Hi there, I just want to say thank's to everyone involved in this PR !
That's a very cool feature. I will test it soon !
